src/relay/pass/partial_eval.cc

```diff
@@ -0,0 +1,764 @@
+    }
+  };
+  StatefulOpVisitor sov;
+  sov(e);
+  return sov.stateful;
+}
+
+using FInterpreter = runtime::TypedPackedFunc<Value(Expr)>;
+
+DLContext CPUContext() {
+  DLContext ctx;
+  ctx.device_type = kDLCPU;
+  ctx.device_id = 0;
+  return ctx;
+}
+
+FInterpreter CPUInterpreter() {
+  Target target = Target::create("llvm");
+  // use a fresh build context
+  // in case we are already in a build context.
+  BuildConfigContext fresh_build_ctx(build_config());
+
+  return CreateInterpreter(Module(nullptr), CPUContext(), target);
+}
+
+class PartialEvaluator : public ExprFunctor<PStatic(const Expr& e, LetList* ll)>,
+                         public PatternFunctor<MatchStatus(const Pattern&, const PStatic&)> {
+ public:
+  PartialEvaluator(const tvm::Array<Var>& free_vars) {
+    for (const Var& v : free_vars) {
+      env_.Insert(v, NoStatic(v));
+    }
+  }
+
+  PStatic VisitExpr_(const ConstantNode* op, LetList* ll) final {
+    return HasStatic(MkSTensor(op->data.CopyTo(context_)), ll->Push(GetRef<Expr>(op)));
+  }
+
+  PStatic VisitExpr_(const TupleNode* op, LetList* ll) final {
+    std::vector<PStatic> value;
+    tvm::Array<Expr> expr;
+    for (const Expr& e : op->fields) {
+      PStatic ps = VisitExpr(e, ll);
+      value.push_back(ps);
+      expr.push_back(ps->dynamic);
+    }
+    return HasStatic(MkSTuple(value), ll->Push(TupleNode::make(expr)));
+  }
+
+  PStatic VisitExpr_(const TupleGetItemNode* op, LetList* ll) final {
+    PStatic ps = VisitExpr(op->tuple, ll);
+    if (ps->pstatic.defined()) {
+      return Downcast<STuple>(ps->pstatic)->fields[op->index];
+    } else {
+      return NoStatic(ll->Push(TupleGetItemNode::make(ps->dynamic, op->index)));
+    }
+  }
+
+  PStatic VisitExpr_(const VarNode* op, LetList* ll) final {
+    return env_.Lookup(GetRef<Var>(op));
+  }
+
+  PStatic VisitExpr_(const GlobalVarNode* op, LetList* ll) final {
+    return NoStatic(GetRef<Expr>(op));
+  }
+
+  PStatic VisitExpr_(const LetNode* op, LetList* ll) final {
+    env_.Insert(op->var, VisitExpr(op->value, ll));
+    return VisitExpr(op->body, ll);
+  }
+
+  PStatic VisitExpr_(const IfNode* op, LetList* ll) final {
+    PStatic c = VisitExpr(op->cond, ll);
+    if (c->pstatic.defined()) {
+      NDArray cpu_array = Downcast<STensor>(c->pstatic)->data.CopyTo(CPUContext());
+      CHECK_EQ(TVMType2Type(cpu_array->dtype), Bool());
+      if (reinterpret_cast<uint8_t*>(cpu_array->data)[0]) {
+        return VisitExpr(op->true_branch, ll);
+      } else {
+        return VisitExpr(op->false_branch, ll);
+      }
+    } else {
+      Expr t = store_.Extend<Expr>([&]() {
+          return LetList::With([&](LetList* ll) {
+              return VisitExpr(op->true_branch, ll)->dynamic;
+            });
+        });
+      Expr f = store_.Extend<Expr>([&]() {
+          return LetList::With([&](LetList* ll) {
+              return VisitExpr(op->false_branch, ll)->dynamic;
+            });
+        });
+      store_.Invalidate();
```

Comment:
Is this `Invalidate` call needed if you are creating a new scope for true/false branch and exit it afterwards via `store.Extend`?

Reply:
Yes.
suppose that both if and else branch are the same, and they contain a write to an unknown reference.

Reply:
Thanks. Could you add a test case? I disabled `Invalidate` and run `test_pass_partial_eval.py`, `test_pass_dead_code_elimination.py`, no test failure. I think it's good to have the test coverage.

Reply:
sure